### PR TITLE
MGMT-7262: Create a ServiceMonitor when deploying Assisted Service

### PR DESCRIPTION
**Pull request by flaper87**
# Assisted Pull Request

## Description

This commit adds, to the operator, the ability to create a ServiceMonitor CR when Assisted Service is being
deployed. This will allow for collecting Assited Service metrics and send them to the local prometheus.

By doing this, it will be possible to have OCM and other tools collect metrics from the local prometheus. It will also
be possible to query prometheus directly, and deploy custom Assisted Service dashboards to grafana

Signed-off-by: Flavio Percoco <flavio@redhat.com>

## List all the issues related to this PR

- [x] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested)
- [ ] No tests needed

Tested this with CRC and following the local-development guidelines in the `operator.md` docs:

1. Built images (assisted-service and bundle)
2. Pushed to `quay.io/fpercoco/$IMAGE:service-monitor`
3. `operator-sdk -n assisted-installer run bundle quay.io/fpercoco/assisted-service-operator-bundle:service-monitor`
4. Modified the CSV to point to my build of assisted-service
5. Created an `AgentServiceConfig`
6. Watched it all roll out 

image

## Assignees

/cc @djzager 

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by djzager** [comment]
visually, this seems legit. Will test it via dev-scripts early next week.

Is there anything more than you testing it manually before you move it out of draft?

File: internal/controller/controllers/agentserviceconfig_controller.go [hunk 1]
```diff
@@ -68,6 +69,7 @@ const (
 
 	defaultIngressCertCMName      string = "default-ingress-cert"
 	defaultIngressCertCMNamespace string = "openshift-config-managed"
+	defaultOpenshiftMonitoringNS  string = "openshift-monitoring"
```
Comment (djzager) [inline internal/controller/controllers/agentserviceconfig_controller.go:72]: unused?

File: internal/controller/controllers/agentserviceconfig_controller.go [hunk 2]
```diff
@@ -169,6 +171,28 @@ func (r *AgentServiceConfigReconciler) Reconcile(origCtx context.Context, req ct
 	return ctrl.Result{}, r.Status().Update(ctx, instance)
 }
 
+func (r *AgentServiceConfigReconciler) ensureServiceMonitor(ctx context.Context, log logrus.FieldLogger, instance *aiv1beta1.AgentServiceConfig) error {
```
Comment (djzager) [inline internal/controller/controllers/agentserviceconfig_controller.go:174]: Add this to the reconcile loop?

**Review by djzager — CHANGES_REQUESTED** [review]
Couple comments.

File: internal/controller/controllers/agentserviceconfig_controller.go (hunk 2 above)
Comment (flaper87) [reply internal/controller/controllers/agentserviceconfig_controller.go:174]: :facepalm:  I laughed out loud! 

I'm getting back to this now and remembering that I didn't seem to see the `ServiceMonitor` being created, which is what I wanted to debug today (and why this was a draft).... now we both know why it was not being created :stuck_out_tongue:

**Review by djzager — APPROVED** [review]

**Pull request merged**